src/boxstream.rs

```diff
@@ -258,33 +288,41 @@ enum RecvState {
 pub struct BoxStreamRecv {
     key_nonce: KeyNonce,
     state: RecvState,
+    goodbye: bool,
 }
 
-// TODO: Handle goodbye
 impl BoxStreamRecv {
     /// Create a new `BoxStreamRecv` from the `key_nonce`.
     pub fn new(key_nonce: KeyNonce) -> Self {
         Self {
             key_nonce,
             state: RecvState::ExpectHeader,
+            goodbye: false,
         }
     }
+
     /// Decrypt a single boxstream message every two calls (one to decrypt and parse the header, the
     /// other do decrypt the body) by decrypting from `buf` and writting the plaintex into `dec`.
-    /// Returns the number of bytes read from `buf` and the number of bytes written into `dec`.
+    /// Returns the number of bytes read from `buf` and the number of bytes written into `dec`.  If
+    /// a goodbye message is received, an Err(Error::Goodbye) will be returned and the following
+    /// calls will return Ok((0, 0)).
     pub fn decrypt(&mut self, buf: &[u8], dec: &mut [u8]) -> Result<(usize, usize)> {
-        // NOTE: Maybe this check should be removed because reading 0 bytes can be understood as
-        // EOF in a Read trait api.
-        let n = self.recv_bytes();
-        if buf.len() < n {
+        if self.goodbye {
```

Comment:
This is not just making the API more complicated? Why not just returning something like `ErrorKind::NotConnected`?